### DISCUSSION TITLE
Need advice on where to start with video downloads

### DISCUSSION THREAD

**Discussion by ShovelJockey** [tokio-rs/axum#3077] [Q&A] [answered]
### Summary

I have created an axum server that remotely manages the starting/shutdown of a motion detection triggered camera. 

My next step is to have the client also be able to download recorded videos from server but am not quite sure how best to get this done with axum. Seems like a reading the file to a stream then returning a response with a stream body similar to some of the examples [here](https://github.com/tokio-rs/axum/discussions/608) would work with the client reading the stream to a file or passing it to video player directly?

But have also seen protocols like HLS being suggested, tower-http functions for serving files alongside vague mentions of using websockets without much specific detail. Ideally looking for something that isn't especially complex while still quite performant that would be flexible enough for a client to be able to save to file or encode and play without needing to save, would really appreciate some advice from someone more experienced with working with media in rust.

These videos are generally quite small in size (between 1-30Mb mostly) and currently am using mp4 format.

### axum version

0.7.5

**Comment by yanns** [comment, marked as answer]
You could use something like https://videojs.com/ in the browser.

On the backend side, you would need to support HTTP range requests support. You could have a look at this: https://github.com/tokio-rs/axum/pull/3047

**Reply by ShovelJockey** [reply]
Thanks for the response.

This is quite interesting to me as it might make building an app for my android phone a little easier if video can be viewed via browser seemed there would be a few hurdles otherwise.

From the the link is this essential for being able to stream video via axum or does it more make it better/easier? Definitely still new to streaming in general and curious.

**Reply by yanns** [reply]
when playing videos, you need to be able to go forward / backward depending on the user interaction.
For some players, you need HTTP range requests: https://en.wikipedia.org/wiki/Byte_serving
See https://stackoverflow.com/questions/13043816/html5-video-and-partial-range-http-requests
Your setup will depend on the video player you'll pick, and maybe on the browser.

**Reply by ShovelJockey** [reply]
Excellent links, the example from the issue worked very well and has been helpful for understanding how things are working under the hood.

Just need to look at managing the stream on the other end instead of testing with curl. Thanks for the help.
